src/vec.rs

```diff
@@ -0,0 +1,364 @@
+            p += Bytes::from(4u64);
+            vec![0; size]
+        };
+        self.memory.read(p.get(), &mut buf);
+        T::from_bytes(buf)
+    }
+
+    #[inline]
+    fn write_value(&mut self, index: usize, value: &T) {
+        let value = value.to_bytes();
+        let mut p = self.index_to_entry_addr(index);
+        if !T::FIXED_SIZE {
+            write_u32(&self.memory, p, value.len() as u32);
+            p += Bytes::from(4u64);
+        }
+        self.memory.write(p.get(), &value);
+    }
+
+    pub fn insert(&mut self, index: usize, element: &T) {
+        let len = self.len();
+        assert!(index <= len);
+
+        // space for the new element
+        if len == self.capacity() {
+            self.reserve(1);
+        }
+
+        // Move all entries starting at index one to the right.
+        let src = self.index_to_entry_addr(index);
+        let dst = self.index_to_entry_addr(index + 1);
+        copy(
+            &self.memory,
+            src,
+            dst,
+            (len - index) as u64,
+            Self::ENTRY_SIZE,
+        );
+
+        self.write_value(index, element);
+
+        self.len += 1;
+        self.save();
+    }
+
+    pub fn remove(&mut self, index: usize) -> T {
+        let len = self.len();
+        assert!(index < len);
+
+        let ret = self.read_value(index);
+
+        // Shift everything down to fill in that spot.
+        copy::<M>(
+            &self.memory,
+            self.index_to_entry_addr(index + 1),
+            self.index_to_entry_addr(index),
+            (len - index - 1) as u64,
+            Self::ENTRY_SIZE,
+        );
+
+        self.len -= 1;
+        self.save();
+        ret
+    }
+
+    #[inline]
+    pub fn push(&mut self, value: &T) {
+        self.write_value(self.len(), value);
+        self.len += 1;
+        self.save();
+    }
+
+    #[inline]
+    pub fn pop(&mut self) -> Option<T> {
+        if self.len == 0 {
+            None
+        } else {
+            self.len -= 1;
+            self.save();
+            Some(self.read_value(self.len()))
+        }
+    }
+
+    /// # Safety
+    ///
+    /// Calling this method with an out-of-bounds index is *[undefined behavior]*
+    /// even if the resulting reference is not used.
+    pub unsafe fn get_unchecked(&self, index: usize) -> T {
+        self.read_value(index)
+    }
+
+    pub fn get(&self, index: usize) -> Option<T> {
+        if index >= self.len() {
+            None
+        } else {
+            Some(self.read_value(index))
+        }
+    }
+
+    pub fn set(&mut self, index: usize, value: &T) {
+        assert!(index < self.len());
+        self.write_value(index, value);
+    }
+
+    pub fn reserve(&mut self, additional: usize) {
+        let page_size = 64 * 1024;
+        let round_up = |x| ((x + page_size - 1) & !(page_size - 1));
+
+        // We need to reserve space for the u32 that indicates the size.
+        let add_bytes = (additional * Self::ENTRY_SIZE) as u64;
```

Comment:
I wonder if we're better off using `leb128` encoding for the size to cut down on the number of bytes needed.